src/java/org/apache/cassandra/gms/Gossiper.java

```diff
@@ -158,38 +158,53 @@
 
     private volatile long lastProcessedMessageAt = System.currentTimeMillis();
 
-    //This property and anything that checks it should be removed in 5.0
-    private boolean haveMajorVersion3Nodes = true;
+    // This property and anything that checks it should be removed in 5.0
+    /**
+     * This property is initially set to {@code true} which means that we have no information about the other nodes.
+     * Once all nodes are on at least this node version, it becomes {@code false}, which means that we are not
+     * upgrading from the previous version (major, minor).
+     */
+    private boolean upgradeInProgressPossible = true;
```

Comment:
We should add `volatile` to the variable, since it is now read by both the supplier and `hasMajorVersion3Nodes()`. It can be accessed from multiple threads. 
Before the patch, the variable is only accessed by `ExpiringMemoizingSupplier`.

Reply:
TBH, I haven't change anything in this regards, the supplier is not asynchronous, but I can make it volatile